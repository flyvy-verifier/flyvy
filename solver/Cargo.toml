[package]
name = "solver"
version.workspace = true
edition.workspace = true

[dependencies]
fly = { path = "../fly" }
smtlib = { path = "../smtlib" }

lazy_static = "1.4.0"
itertools = "0.10.5"
serde = { version = "1.0.164", features = ["derive"] }
serde_derive = "1.0.164"
thiserror = "1.0.40"
regex = "1.8.4"
pretty_env_logger = "0.5.0"
log = "0.4.19"
rayon = "1.7.0"
peg = "0.8.1"

[dev-dependencies]
test-log = "0.2.11"
env_logger = "0.10.0"
