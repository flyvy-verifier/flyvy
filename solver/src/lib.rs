// Copyright 2022-2023 VMware, Inc.
// SPDX-License-Identifier: BSD-2-Clause

//! High-level interface to an SMT solver using [`Term`](crate::fly::syntax::Term).

// configure clippy
#![allow(clippy::needless_return)]
#![allow(clippy::large_enum_variant)]
#![allow(clippy::upper_case_acronyms)]
#![allow(clippy::type_complexity)]
// documentation-related lints (only checked when running rustdoc)
#![warn(missing_docs)]
#![allow(rustdoc::private_intra_doc_links)]
#![deny(rustdoc::broken_intra_doc_links)]

pub mod backends;
pub mod conf;
pub mod imp;
pub mod models;
pub mod sexp;

pub use smtlib::path::{log_dir, solver_path};
pub use smtlib::proc::{SatResp, SmtPid};
