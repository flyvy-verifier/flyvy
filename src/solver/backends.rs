// Copyright 2022-2023 VMware, Inc.
// SPDX-License-Identifier: BSD-2-Clause

//! Support for launching a solver (Z3, CVC4, or CVC5) and then parsing its
//! models, which are the two features that generally differ from solver to
//! solver.

use std::{
    collections::{HashMap, HashSet},
    iter::zip,
    time::Instant,
};

use crate::{
    fly::{
        semantics::{Element, Interpretation},
        syntax::{Signature, Sort},
    },
    smtlib::{
        proc::{CvcConf, SolverCmd, Z3Conf},
        sexp::{self, Atom},
    },
    solver::models::ModelSymbol,
};

use super::{
    models::{self, PartialInterp},
    {Backend, FOModel},
};

/// The type of solver being used
#[allow(missing_docs)]
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum SolverType {
    Z3,
    Cvc4,
    Cvc5,
}

/// A Backend for launching and parsing Z3/CVC4/CVC5, with some hard-coded options.
#[derive(Debug, Clone)]
pub struct GenericBackend {
    solver_type: SolverType,
    bin: String,
}

impl GenericBackend {
    /// Create a Backend for a given type of solver and with a path to the
    /// solver binary.
    pub fn new(solver_type: SolverType, bin: &str) -> Self {
        Self {
            solver_type,
            bin: bin.to_string(),
        }
    }
}

fn sort_cardinality(universes: &HashMap<String, usize>, sort: &Sort) -> usize {
    match sort {
        Sort::Bool => 2,
        Sort::Id(s) => *universes
            .get(s)
            .unwrap_or_else(|| panic!("unknown sort {s}")),
    }
}

impl Backend for &GenericBackend {
    fn get_cmd(&self) -> SolverCmd {
        match self.solver_type {
            SolverType::Z3 => {
                let mut conf = Z3Conf::new(&self.bin);
                conf.model_compact();
                conf.done()
            }
            SolverType::Cvc4 => {
                let mut conf = CvcConf::new_cvc4(&self.bin);
                conf.finite_models();
                conf.interleave_enumerative_instantiation();
                conf.done()
            }
            SolverType::Cvc5 => {
                let mut conf = CvcConf::new_cvc5(&self.bin);
                conf.finite_models();
                conf.interleave_enumerative_instantiation();
                conf.done()
            }
        }
    }

    fn parse(
        &self,
        sig: &Signature,
        // TODO(tej): unused, maybe we can remove from Backend trait
        _n_states: usize,
        indicators: &HashSet<String>,
        model: &sexp::Sexp,
    ) -> FOModel {
        let model = match self.solver_type {
            SolverType::Z3 => models::parse_z3(model),
            SolverType::Cvc4 => models::parse_cvc(model, false),
            SolverType::Cvc5 => models::parse_cvc(model, true),
        };

        let universe: HashMap<String, usize> = model
            .universes
            .iter()
            .map(|(sort, elements)| (sort.clone(), elements.len()))
            .collect();

        let mut part_interp = PartialInterp::for_model(&model);
        let mut symbols = model.symbols.iter().collect::<Vec<_>>();
        symbols.sort_unstable_by_key(|(symbol, _)| {
            let in_signature = sig.contains_name(symbol);
            // sort the model-only signatures first
            //
            // NOTE: this is just a heuristic to evaluate auxilliary functions
            // first, for now.
            (in_signature, symbol.to_string())
        });
        for (symbol, model_sym) in symbols.into_iter() {
            if indicators.contains(symbol) {
                continue;
            }
            let start = Instant::now();
            let ModelSymbol {
                binders,
                body,
                ret_sort,
            } = model_sym;
            let arg_sorts = binders
                .iter()
                .map(|(_, sort)| sort.clone())
                .collect::<Vec<_>>();
            let mut shape = arg_sorts
                .iter()
                .map(|sort| sort_cardinality(&universe, sort))
                .collect::<Vec<usize>>();
            shape.push(sort_cardinality(&universe, ret_sort));
            let interp = Interpretation::new(&shape, |args: &[Element]| -> Element {
                // get the arguments as terms, based on model.universes
                let args = zip(args, &arg_sorts)
                    .map(|(&e_idx, typ)| match typ {
                        Sort::Bool => {
                            if e_idx == 0 {
                                Atom::S("false".to_string())
                            } else {
                                Atom::S("true".to_string())
                            }
                        }
                        Sort::Id(sort) => {
                            let elements = &model.universes[sort];
                            let element = elements[e_idx].clone();
                            Atom::S(element)
                        }
                    })
                    .collect::<Vec<_>>();
                let repl: HashMap<&str, Atom> =
                    zip(binders.iter().map(|s| s.0.as_str()), args).collect();
                let e = model
                    .smt_eval(&repl, &part_interp, body)
                    .unwrap_or_else(|err| panic!("could not interpret {symbol}: {err}"));
                let res = e.s().expect("unhandled evaluation result");
                match ret_sort {
                    Sort::Bool => {
                        if res == "false" {
                            return 0;
                        } else if res == "true" {
                            return 1;
                        } else {
                            panic!("unexpected bool {res}")
                        }
                    }
                    Sort::Id(sort) => {
                        let elements = &model.universes[sort];
                        let res_idx = elements
                            .iter()
                            .position(|x| x == res)
                            .unwrap_or_else(|| panic!("unknown {sort} element {res}"));
                        return res_idx as Element;
                    }
                }
            });
            part_interp
                .interps
                .insert(symbol.clone(), (interp, ret_sort.clone()));
            log::debug!(
                "evaluated {symbol} in {:.1}s",
                Instant::now().duration_since(start).as_secs_f64()
            )
        }

        let interp = part_interp
            .interps
            .into_iter()
            .filter(|(symbol, _)| sig.contains_name(symbol))
            .map(|(symbol, (interp, _))| (symbol, interp))
            .collect();
        FOModel { universe, interp }
    }
}

#[cfg(test)]
mod tests {
    use std::{
        collections::{HashMap, HashSet},
        fs,
    };

    use crate::{
        fly::{parser, syntax::Term},
        smtlib::{proc::SatResp, sexp},
        solver::{solver_path, Solver},
    };

    use super::{Backend, GenericBackend, SolverType};

    use test_log::test;

    #[test]
    fn test_issue_5_parse_model_with_auxilliary_defs() {
        let _ = pretty_env_logger::try_init();
        let sig = parser::parse_signature(
            r#"
        sort node
        sort quorum
        sort value

        # constants:


        # functions:


        # relations:
        mutable member(node, quorum): bool
        mutable vote_request_msg(node, node): bool
        mutable voted(node): bool
        mutable vote_msg(node, node): bool
        mutable votes(node, node): bool
        mutable leader(node): bool
        mutable decided(node, value): bool
        "#,
        );

        let backend = GenericBackend {
            solver_type: SolverType::Z3,
            bin: "z3".to_string(),
        };

        let model_text =
            fs::read_to_string("tests/issue_5_model.sexp").expect("could not find model file");
        let model_sexp = sexp::parse(&model_text).expect("test model does not parse");

        let fo_model = (&backend).parse(&sig, 1, &HashSet::new(), &model_sexp);
        // a (primed) relation from the signature
        assert!(fo_model.interp.contains_key("leader'"));
        // auxilliary definition in Z3's model
        assert!(!fo_model.interp.contains_key("k!1058"));
    }

    #[test]
    fn test_parse_test_model() {
        let _ = pretty_env_logger::try_init();
        let sig = parser::parse_signature(
            r#"
        sort node
        mutable votes(node, node): bool
        "#,
        );

        let backend = GenericBackend {
            solver_type: SolverType::Z3,
            bin: solver_path("z3"),
        };

        let model_text =
            fs::read_to_string("tests/test_model.sexp").expect("could not find model file");
        let model_sexp = sexp::parse(&model_text).expect("test model does not parse");

        let fo_model = (&backend).parse(&sig, 0, &HashSet::new(), &model_sexp);
        assert!(fo_model.interp.contains_key("votes"));
    }

    #[test]
    fn test_check_sat_assuming_and_get_minimal_model() {
        let sig = parser::parse_signature(
            r#"
            sort A
            mutable x: A
            mutable p(A): bool
        "#,
        );
        let backend = GenericBackend {
            solver_type: SolverType::Z3,
            bin: solver_path("z3"),
        };
        let mut solver =
            Solver::new(&sig, 1, &backend, None).expect("could not create solver for test");
        let ind = solver.get_indicator("i");
        // make sure that the first universe tried is not minimal
        solver.assert(&parser::term("exists a1:A, a2:A. a1 != a2"));
        solver.assert(&Term::implies(ind.clone(), parser::term("p(x)")));
        let mut assumptions = HashMap::new();
        assumptions.insert(ind, true);
        let resp = solver.check_sat(assumptions).unwrap();
        assert!(resp == SatResp::Sat);
        let model = &solver.get_minimal_model()[0];
        assert_eq!(
            model.eval(&parser::term("p(x)")),
            1,
            "p(x) should be true due to assumption"
        );
    }
}
