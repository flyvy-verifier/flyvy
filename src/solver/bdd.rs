// Copyright 2022-2023 VMware, Inc.
// SPDX-License-Identifier: BSD-2-Clause

//! A bounded model checker for flyvy programs using symbolic evaluation.

use crate::fly::{sorts::*, syntax::*};
use crate::term::FirstOrder;
use biodivine_lib_bdd::*;
use itertools::Itertools;
use std::collections::HashMap;
use thiserror::Error;

/// Holds an ordering of all (relation, elements) pairs
#[allow(dead_code)]
struct Context<'a> {
    signature: &'a Signature,
    universe: &'a Universe,

    len: usize,
    idxs: HashMap<&'a str, HashMap<Vec<usize>, usize>>,

    bdds: BddVariableSet,
    vars: Vec<BddVariable>,
}

impl Context<'_> {
    fn new<'a>(signature: &'a Signature, universe: &'a Universe) -> Context<'a> {
        let order = signature.relations.iter().flat_map(|relation| {
            if relation.args.is_empty() {
                vec![(relation.name.as_str(), (vec![]))]
            } else {
                relation
                    .args
                    .iter()
                    .map(|sort| cardinality(universe, sort))
                    .map(|card| (0..card).collect::<Vec<usize>>())
                    .multi_cartesian_product()
                    .map(|element| (relation.name.as_str(), element))
                    .collect()
            }
        });

        let mut len = 0;
        let mut idxs: HashMap<_, HashMap<_, _>> = HashMap::new();
        for (i, (r, e)) in order.enumerate() {
            len += 1;
            idxs.entry(r).or_default().insert(e, i);
        }

        let bdds = BddVariableSet::new_anonymous((idxs.len() * 2).try_into().unwrap());
        let vars = bdds.variables();

        Context {
            signature,
            universe,
            len,
            idxs,
            bdds,
            vars,
        }
    }

    fn get(&self, relation: &str, element: &[usize], prime: bool) -> Bdd {
        self.bdds
            .mk_var(self.vars[self.idxs[relation][element] + if prime { self.len } else { 0 }])
    }
}

/// The result of a successful run of the bounded model checker
#[derive(Debug)]
pub enum CheckerResult {
    /// The checker found a counterexample
    Counterexample,
    /// The checker did not find a counterexample
    Unknown,
}

#[allow(missing_docs)]
#[derive(Error, Debug)]
pub enum CheckerError {
    #[error("sort checking error: {0}")]
    SortError(SortError),
    #[error("sort {0} not found in universe {1:#?}")]
    UnknownSort(String, Universe),

    #[error("all assumes should precede all asserts, but found {0:?}")]
    OutOfOrderStatement(ThmStmt),

    #[error("expected no primes in {0}")]
    AnyFuture(Term),
    #[error("expected no primes or only one prime in {0}")]
    TooFuture(Term),
    #[error("expected all asserts be safety properties but found {0}")]
    AssertWithoutAlways(Term),

    #[error("expected a boolean expression but found an element")]
    ExpectedBdd,
    #[error("expected an element of a sort but found an expression {0:?}")]
    ExpectedElement(Bdd),
    #[error("could not translate term {0}")]
    UnsupportedTerm(Term),
}

/// Map from uninterpreted sort names to sizes
type Universe = HashMap<String, usize>;

fn cardinality(universe: &Universe, sort: &Sort) -> usize {
    match sort {
        Sort::Bool => 2,
        Sort::Id(id) => universe[id],
    }
}

/// Check a given Module out to some depth
pub fn check(
    module: &mut Module,
    universe: &Universe,
    _depth: usize,
) -> Result<CheckerResult, CheckerError> {
    if let Err((error, _)) = sort_check_and_infer(module) {
        return Err(CheckerError::SortError(error));
    }

    for sort in &module.signature.sorts {
        if !universe.contains_key(sort) {
            return Err(CheckerError::UnknownSort(sort.clone(), universe.clone()));
        }
    }

    for relation in &module.signature.relations {
        if relation.sort != Sort::Bool {
            todo!("non-bool relations")
        }
    }

    if !module.defs.is_empty() {
        todo!("definitions are not supported yet");
    }

    let mut assumes = Vec::new();
    let mut asserts = Vec::new();
    for statement in &module.statements {
        match statement {
            ThmStmt::Assert(Proof { assert, invariants }) => {
                asserts.push(assert.x.clone());
                if !invariants.is_empty() {
                    eprintln!("note: invariants are not yet supported, and do nothing")
                }
            }
            ThmStmt::Assume(term) if asserts.is_empty() => assumes.push(term.clone()),
            _ => return Err(CheckerError::OutOfOrderStatement(statement.clone())),
        }
    }

    let mut inits = Vec::new();
    let mut trs = Vec::new();
    for assume in assumes {
        match assume {
            Term::UnaryOp(UOp::Always, axiom) if FirstOrder::unrolling(&axiom) == Some(0) => {
                inits.push(*axiom.clone());
                trs.push(*axiom);
            }
            Term::UnaryOp(UOp::Always, tr) if FirstOrder::unrolling(&tr) == Some(1) => {
                trs.push(*tr)
            }
            Term::UnaryOp(UOp::Always, term) => return Err(CheckerError::TooFuture(*term)),
            init if FirstOrder::unrolling(&init) == Some(0) => inits.push(init),
            init => return Err(CheckerError::AnyFuture(init)),
        }
    }

    let mut safes = Vec::new();
    for assert in asserts {
        match assert {
            Term::UnaryOp(UOp::Always, safe) if FirstOrder::unrolling(&safe) == Some(0) => {
                safes.push(*safe)
            }
            Term::UnaryOp(UOp::Always, safe) => return Err(CheckerError::AnyFuture(*safe)),
            assert => return Err(CheckerError::AssertWithoutAlways(assert)),
        }
    }

    let context = Context::new(&module.signature, universe);

    let translate = |terms| {
        let term = Term::NAryOp(NOp::And, terms);
        let term = nullary_id_to_app(term, &module.signature.relations);
        let term = crate::term::Next::new(&module.signature).normalize(&term);
        term_to_bdd(&term, &context, &HashMap::new())?.bdd()
    };
    let _init = translate(inits)?;
    let _tr = translate(trs)?;
    let _safe = translate(safes)?;

    todo!()
}

fn nullary_id_to_app(term: Term, relations: &[RelationDecl]) -> Term {
    match term {
        Term::Id(id) if relations.iter().any(|r| r.name == id) => Term::App(id, 0, vec![]),
        Term::App(name, primes, args) => Term::App(
            name,
            primes,
            args.into_iter()
                .map(|arg| nullary_id_to_app(arg, relations))
                .collect(),
        ),
        Term::UnaryOp(op, term) => Term::UnaryOp(op, Box::new(nullary_id_to_app(*term, relations))),
        Term::BinOp(op, a, b) => Term::BinOp(
            op,
            Box::new(nullary_id_to_app(*a, relations)),
            Box::new(nullary_id_to_app(*b, relations)),
        ),
        Term::NAryOp(op, args) => Term::NAryOp(
            op,
            args.into_iter()
                .map(|arg| nullary_id_to_app(arg, relations))
                .collect(),
        ),
        Term::Ite { cond, then, else_ } => Term::Ite {
            cond: Box::new(nullary_id_to_app(*cond, relations)),
            then: Box::new(nullary_id_to_app(*then, relations)),
            else_: Box::new(nullary_id_to_app(*else_, relations)),
        },
        Term::Quantified {
            quantifier,
            binders,
            body,
        } => Term::Quantified {
            quantifier,
            binders,
            body: Box::new(nullary_id_to_app(*body, relations)),
        },
        term => term,
    }
}

#[derive(Debug)]
enum BddOrElement {
    Bdd(Bdd),
    Element(usize),
}

impl BddOrElement {
    fn bdd(self) -> Result<Bdd, CheckerError> {
        match self {
            BddOrElement::Bdd(bdd) => Ok(bdd),
            BddOrElement::Element(_) => Err(CheckerError::ExpectedBdd),
        }
    }

    fn element(self) -> Result<usize, CheckerError> {
        match self {
            BddOrElement::Element(element) => Ok(element),
            BddOrElement::Bdd(bdd) => Err(CheckerError::ExpectedElement(bdd)),
        }
    }
}

fn term_to_bdd(
    term: &Term,
    context: &Context,
    assignments: &HashMap<String, usize>,
) -> Result<BddOrElement, CheckerError> {
    let go = |term| term_to_bdd(term, context, assignments);

    let bdd: Bdd = match term {
        Term::Literal(true) => context.bdds.mk_true(),
        Term::Literal(false) => context.bdds.mk_false(),
        Term::Id(id) => return Ok(BddOrElement::Element(assignments[id])),
        Term::App(relation, primes, args) => context.get(
            relation,
            &args
                .iter()
                .map(|arg| go(arg)?.element())
                .collect::<Result<Vec<_>, _>>()?,
            *primes == 1,
        ),
        Term::UnaryOp(UOp::Not, term) => go(term)?.bdd()?.not(),
        Term::BinOp(BinOp::Iff, a, b) => go(a)?.bdd()?.iff(&go(b)?.bdd()?),
        Term::BinOp(BinOp::Equals, a, b) => match (go(a)?, go(b)?) {
            (BddOrElement::Bdd(a), BddOrElement::Bdd(b)) => a.iff(&b),
            (BddOrElement::Element(a), BddOrElement::Element(b)) if a == b => {
                context.bdds.mk_true()
            }
            (BddOrElement::Element(a), BddOrElement::Element(b)) if a == b => {
                context.bdds.mk_false()
            }
            _ => return Err(CheckerError::UnsupportedTerm(term.clone())),
        },
        Term::BinOp(BinOp::NotEquals, a, b) => {
            go(&Term::BinOp(BinOp::Equals, a.clone(), b.clone()))?
                .bdd()?
                .not()
        }
        Term::BinOp(BinOp::Implies, a, b) => go(a)?.bdd()?.imp(&go(b)?.bdd()?),
        Term::NAryOp(NOp::And, terms) => {
            terms.iter().fold(Ok(context.bdds.mk_true()), |and, term| {
                Ok(and?.and(&go(term)?.bdd()?))
            })?
        }
        Term::NAryOp(NOp::Or, terms) => {
            terms.iter().fold(Ok(context.bdds.mk_false()), |or, term| {
                Ok(or?.or(&go(term)?.bdd()?))
            })?
        }
        Term::Quantified {
            quantifier,
            binders,
            body,
        } => {
            assert!(!binders.is_empty());
            let shape: Vec<usize> = binders
                .iter()
                .map(|b| cardinality(context.universe, &b.sort))
                .collect();
            let names: Vec<&String> = binders.iter().map(|b| &b.name).collect();
            let bodies = shape
                .iter()
                .map(|&card| (0..card).collect::<Vec<usize>>())
                .multi_cartesian_product()
                .map(|elements| {
                    // extend assignments with all variables bound to these `elements`
                    let mut assignments = assignments.clone();
                    for (name, element) in names.iter().zip(elements) {
                        assignments.insert(name.to_string(), element);
                    }
                    go(body)?.bdd()
                })
                .collect::<Result<Vec<_>, _>>()?;
            match quantifier {
                Quantifier::Forall => bodies
                    .into_iter()
                    .fold(context.bdds.mk_true(), |and, term| and.and(&term)),
                Quantifier::Exists => bodies
                    .into_iter()
                    .fold(context.bdds.mk_false(), |or, term| or.or(&term)),
            }
        }
        _ => return Err(CheckerError::UnsupportedTerm(term.clone())),
    };
    Ok(BddOrElement::Bdd(bdd))
}
