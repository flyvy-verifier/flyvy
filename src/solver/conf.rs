use std::path::PathBuf;

use crate::fly::syntax::Signature;

use super::{backends::GenericBackend, Solver};

/// Wrapper around the configuration needed to launch a solver.
#[derive(Debug, Clone)]
pub struct SolverConf {
    /// Which backend to use for launched solvers.
    pub backend: GenericBackend,
    /// The optional path to tee SMT output to.
    pub tee: Option<PathBuf>,
}

impl SolverConf {
    /// Launch a new solver with the given configuration.
    pub fn solver(&self, sig: &Signature, n_states: usize) -> Solver<&GenericBackend> {
        // TODO: failures to start the solver should be bubbled up to user nicely
        Solver::new(sig, n_states, &self.backend, self.tee.as_deref())
            .expect("could not start solver")
    }
}
