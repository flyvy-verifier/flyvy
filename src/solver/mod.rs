// Copyright 2022-2023 VMware, Inc.
// SPDX-License-Identifier: BSD-2-Clause

//! High-level interface to an SMT solver using [`Term`](crate::fly::syntax::Term).

pub mod backends;
mod imp;
mod models;
mod path;
mod sexp;
pub use imp::{Backend, FOModel, Solver};
pub use path::solver_path;
mod conf;
pub use conf::SolverConf;
