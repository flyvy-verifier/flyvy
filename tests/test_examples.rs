// Copyright 2022-2023 VMware, Inc.
// SPDX-License-Identifier: BSD-2-Clause

#![allow(clippy::needless_return)]

use std::{
    collections::HashMap,
    env,
    ffi::OsStr,
    fmt::Display,
    fs::{self, File},
    io::{self, BufRead},
    path::{Path, PathBuf},
    process::Command,
};

use clap::Parser;
use lazy_static::lazy_static;
use regex::Regex;
use serde_derive::Deserialize;
use temporal_verifier::solver::solver_path;
use walkdir::WalkDir;

const SOLVERS_TO_TEST: [&str; 3] = ["z3", "cvc4", "cvc5"];

lazy_static! {
    static ref SOLVER_VERSION_VARS: HashMap<String, String> = {
        let mut versions = HashMap::new();
        let re = Regex::new(r#"(.*)_VERSION="(.*)""#).unwrap();
        for line in fs::read_to_string("tools/solver-versions.sh")
            .expect("could not find tools/solver-versions.sh")
            .lines()
        {
            if line.starts_with('#') || line.is_empty() {
                continue;
            }
            let m = re
                .captures(line)
                .expect("malformed line in solver-versions.sh");
            let solver = m.get(1).unwrap().as_str();
            let version = m.get(2).unwrap().as_str();
            versions.insert(solver.to_string(), version.to_string());
        }
        versions
    };
}

fn expected_version(solver: &str) -> String {
    let version = SOLVER_VERSION_VARS
        .get(&solver.to_uppercase())
        .expect("unexpected solver {solver} (not in tools/solver-versions.sh)");
    if solver == "z3" {
        format!("Z3 version {version}")
    } else if solver == "cvc4" {
        format!("CVC4 version {version}")
    } else if solver == "cvc5" {
        format!("cvc5 version {version}")
    } else {
        panic!("unexpected solver {solver}")
    }
}

fn get_version(solver: &str) -> String {
    let bin = solver_path(solver);
    let version_out = Command::new(bin).arg("--version").output();
    let version = match version_out {
        Ok(out) => String::from_utf8(out.stdout)
            .unwrap()
            .lines()
            .next()
            .expect("no output from --version")
            .to_string(),
        Err(_) => return format!("({solver} not found)"),
    };
    let version_re = Regex::new(r"(Z3|CVC4|cvc5) version [0-9.]*").unwrap();
    match version_re.captures(&version) {
        Some(capture) => capture.get(0).unwrap().as_str().to_string(),
        None => panic!("unexpected --version output"),
    }
}

/// Returns true if versions match. On version mismatch, prints the mismatch to
/// stderr for debugging.
fn check_version(solver: &str) -> bool {
    let expected = expected_version(solver);
    let actual = get_version(solver);
    if let Some(s) = env::var_os("TEST_STRICT_VERSIONS") {
        if s == OsStr::new("true") {
            // cause the test to fail
            assert_eq!(
                expected, actual,
                "installed {solver} version does not match"
            );
        }
    }
    if expected != actual {
        eprintln!("solver version mismatch:");
        eprintln!("  expected: {expected}");
        eprintln!("  actual: {actual}");
    }
    return expected == actual;
}

#[derive(Deserialize, Debug, Clone, clap::Parser)]
struct TestCfg {
    /// Assert that verification fails.
    #[arg(long)]
    #[serde(default)]
    expect_fail: bool,

    /// Automatically run this test with all SMT solvers.
    #[arg(long)]
    #[serde(default)]
    all_solvers: bool,

    /// Give this test a name (used in the snapshot file name)
    #[arg(long)]
    name: Option<String>,

    /// Arguments to be passed to the verifier.
    #[arg(last = true)]
    args: Vec<String>,
}

impl TestCfg {
    fn append_to_name(&mut self, s: &str) {
        self.name = match &self.name {
            Some(name) => Some(format!("{name}.{s}")),
            None => Some(s.to_string()),
        };
    }

    fn solver(&self) -> String {
        for arg in &self.args {
            if let Some(solver) = arg.strip_prefix("--solver=") {
                return solver.to_string();
            }
        }
        "z3".to_string()
    }
}

impl Display for TestCfg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if self.expect_fail {
            write!(f, "--expect-fail ")?;
        }
        if self.all_solvers {
            write!(f, "--all-solvers ")?;
        }
        if let Some(name) = &self.name {
            write!(f, "--name={name} ")?;
        }
        if !self.args.is_empty() {
            write!(f, "-- ")?;
            let args = self
                .args
                .iter()
                .map(|arg| shell_words::quote(arg))
                .collect::<Vec<_>>();
            write!(f, "{}", args.join(" "))?;
        }
        Ok(())
    }
}

#[derive(Deserialize, Debug, Clone)]
struct Tests {
    /// Tests to run for every fly file in the config file's directory.
    tests: Vec<TestCfg>,
}

#[derive(Debug, Clone)]
struct Test {
    path: PathBuf,
    cfg: TestCfg,
}

impl Display for Test {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} {}", &self.cfg, self.path.display())
    }
}

/// Get tests specified as TEST lines
fn get_file_tests(path: &Path) -> Vec<Test> {
    let f = File::open(path).expect("could not open test file");
    let f = io::BufReader::new(f);
    f.lines()
        .flat_map(|l| {
            let l = l.unwrap();
            l.strip_prefix("# TEST ").map(|test_line| {
                let split_line = match shell_words::split(test_line) {
                    Ok(split) => split,
                    Err(err) => {
                        eprintln!("could not tokenize TEST line:");
                        eprintln!("# TEST {test_line}");
                        panic!("{err}");
                    }
                };
                let args = ["TEST".to_string()]
                    .into_iter()
                    .chain(split_line.into_iter());
                Test {
                    path: path.to_path_buf(),
                    cfg: TestCfg::try_parse_from(args).unwrap_or_else(|err| {
                        eprintln!("could not parse TEST line:");
                        eprintln!("# TEST {test_line}");
                        panic!("could not parse TEST line: {err}");
                    }),
                }
            })
        })
        .collect()
}

fn get_toml_tests(toml_file: &Path) -> Vec<Test> {
    let f = fs::read_to_string(toml_file).expect("could not open config file");
    let tests: Tests =
        toml::from_str(&f).unwrap_or_else(|err| panic!("could not parse toml file: {err}"));
    let dir = toml_file.parent().unwrap();
    dir.read_dir()
        .expect("could not read toml dir")
        .filter_map(|e| e.ok())
        .filter(|entry| entry.path().extension() == Some(OsStr::new("fly")))
        .flat_map(|entry| {
            tests
                .tests
                .iter()
                .map(|cfg| Test {
                    path: entry.path(),
                    cfg: cfg.clone(),
                })
                .collect::<Vec<_>>()
        })
        .collect()
}

/// Get tests recursively under a root directory, keyed by their file path.
fn get_tests(root_dir: &str) -> HashMap<PathBuf, Vec<Test>> {
    let root_dir = Path::new(root_dir);
    let tests = WalkDir::new(root_dir)
        .sort_by_file_name()
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|entry| entry.file_type().is_file())
        .flat_map(|entry| {
            if entry.path().ends_with("tests.toml") {
                get_toml_tests(entry.path())
            } else if entry.path().extension() == Some(OsStr::new("fly")) {
                get_file_tests(entry.path())
            } else {
                vec![]
            }
        });
    let mut path_tests = HashMap::new();
    for test in tests {
        path_tests
            .entry(test.path.clone())
            .or_insert_with(Vec::new)
            .push(test);
    }
    path_tests
}

fn verifier() -> Command {
    let mut cmd = Command::new("./target/debug/temporal-verifier");
    cmd.arg("--color=never");
    cmd
}

impl Test {
    fn test_name(&self) -> String {
        let path = self.path.file_name().unwrap().to_string_lossy();
        if let Some(name) = &self.cfg.name {
            format!("{path}.{name}")
        } else {
            path.to_string()
        }
    }

    /// Turn all_solvers=true into separate Test structs.
    fn normalize(&self) -> Vec<Self> {
        if self.cfg.all_solvers {
            SOLVERS_TO_TEST
                .map(|solver| {
                    let mut test = self.clone();
                    test.cfg.all_solvers = false;
                    test.cfg.args.push(format!("--solver={solver}"));
                    test.cfg.append_to_name(solver);
                    test
                })
                .to_vec()
        } else {
            vec![self.clone()]
        }
    }

    fn run(&self) {
        if self.cfg.all_solvers {
            for t in self.normalize() {
                t.run()
            }
        } else {
            let out = verifier()
                .args(&self.cfg.args)
                .arg(&self.path)
                .output()
                .expect("could not run verifier");
            let stdout = String::from_utf8(out.stdout).expect("non-utf8 output");
            let stderr = String::from_utf8(out.stderr).expect("non-utf8 output");
            let combined_stdout_stderr =
                format!("{stdout}\n======== STDERR: ===========\n{stderr}");

            if check_version(&self.cfg.solver()) {
                insta::assert_display_snapshot!(self.test_name(), combined_stdout_stderr);
            }

            if self.cfg.expect_fail {
                assert!(
                    !out.status.success(),
                    "verifier succeeded but expected failure"
                );
            } else if !out.status.success() {
                eprintln!("{stderr}");
                assert!(out.status.success(), "verifier should succeed");
            }
        }
    }
}

fn test_dir(root_dir: &str) {
    for (_, tests) in get_tests(root_dir) {
        for (n, mut test) in tests.into_iter().enumerate() {
            // ensure test names are unique by appending an index
            if n >= 1 {
                test.cfg.append_to_name(&format!("{n}"));
            }
            println!("# TEST {test}");
            let test_path = test.path.parent().unwrap().strip_prefix(root_dir).unwrap();
            insta::with_settings!(
                {
                    description => format!("{test}"),
                    prepend_module_to_snapshot => false,
                    // The initial .. is needed because this path is relative to
                    // where this test is
                    snapshot_path => Path::new("..").join(root_dir).join("snapshots").join(test_path),
                },
                { test.run(); }
            )
        }
    }
}

#[test]
fn test_solver_versions_match() {
    let mismatches = SOLVERS_TO_TEST
        .iter()
        .filter(|solver| !check_version(solver))
        .map(|s| s.to_string())
        .collect::<Vec<_>>();
    assert!(
        mismatches.is_empty(),
        "some solvers ({}) are at the wrong version (try running ./tools/download-solvers.sh)",
        mismatches.join(", ")
    );
}

#[test]
fn test_small_examples() {
    test_dir("tests/examples")
}

#[test]
fn test_larger_examples() {
    test_dir("examples")
}
