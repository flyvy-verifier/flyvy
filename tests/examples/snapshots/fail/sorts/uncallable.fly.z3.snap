---
source: tests/test_examples.rs
description: "--expect-fail --all-solvers -- verify tests/examples/fail/sorts/uncallable.fly"
expression: combined_stdout_stderr
---

======== STDERR: ===========
sort checking error:
error: x was called but didn't take any args
  ┌─ tests/examples/fail/sorts/uncallable.fly:7:1
  │
7 │ assert forall x:s. x(true) = x
  │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


